website/content/docs/rfcs/external-keys.mdx

```diff
@@ -0,0 +1,214 @@
+     for any given key using this provider.
+
+     A configuration may be inherited from a direct parent namespace via the
+     special `inherits = "<:config-name>"` key. This inheritance mechanism
+     is chosen over a top-level `inheritable = true` flag to avoid potential
+     naming conflicts across the namespace hierarchy, or alternatively to avoid
+     the need for a mechanism/format to address a specific ancestor namespace.
+     Inherited configurations are similar to symbolic links; they do not copy
+     information to the child namespace or enable read/write access to the
+     original configuration in any way. To propagate configurations across
+     several levels of child namespaces, you would build a chain of inherited
+     configurations.
+
+   - `configs/<:config-name>/keys/<:key-name>`, to maintain mappings of keys
+     and their access information. A `key-name` is unique per `config-name`, not
+     per namespace.
+
+   - `configs/<:config-name>/keys/<:key-name>/grants/<:mount-path>` will be
+     used to add or remove mounts from accessing the key.
+
+     Grants handle both mount paths local to the key configuration's namespace
+     just as mount paths of any child namespaces, e.g. `my-child/pki`. Combined
+     with inherited configurations via `inherits = ...`, this lets a namespace
+     pass down a key to select mounts in child namespaces. Note that grants
+     can only be configured at the "origin" configuration and cannot be set for
+     inherited configurations.
+
+Notably, unlike Vault Enterprise, no key material is implicitly created at
+the key association step; it must already exist within the KMS or HSM and this
+is just a linking.
+
+Furthermore, rather than relying on mount tuning, the use of explicit grants,
+with mounts in the path, allows for fine-grained delegation of permission via
+the standard ACL models, assuming the operator does not have policy
+modification permissions.
+
+In the future, key creation may be supported as a follow-up RFC.
+
+Individual mount types (PKI, SSH, Identity, ...) will need to be updated to
+support "creating" keys via using external keys instead of existing keys. This
+will be done with the `config-name` plus `key-name` association.
+
+## Technical Description
+
+This requires an improvement to the configuration, along with allowing these new
+stanzas to be reloaded via `SIGHUP`.
+
+### SystemView Changes
+
+However, elided from the above user-facing description is the core technical
+improvement: plugins interact with core via the `logical.SystemView`
+[interface](https://github.com/openbao/openbao/blob/48e7bd6f8358191056bdd0d2ac8b11e9e253254d/sdk/logical/system_view.go#L18-L99),
+which will need to be extended to support the following new APIs:
+
+1. `ListKeys() (map[string]*ExternalKeyInfo, error)` - to return information
+   about external keys accessible to this mount.
+2. `GetKey() (go-kms-wrapper.ExternalKey, error)` - to return a helper to
```

Comment:
```suggestion
2. `GetKey() (logical.ExternalKey, error)` - to return a helper to
```

I think wrapping the g-k-w type with an internal type would be good. Would let us implement it more easily in the GRPC layer.

Reply:
> The only advantage I can see for having the interface in api or sdk is that plugins don't need to concern themselves with importing g-k-w to reference the ExternalKey type.

Right, it is a rather fat library so I'd like to avoid it if we can. Also it is somewhat unusual, most of the other pieces of logical live there so I'd prefer to keep it that way.